### PR TITLE
8 data persistance

### PR DESCRIPTION
**Pull request by MadhavKanna**
Description: This pull request implements data persistence functionality and for database in TestDatabaseReadWrite class, and tests for the database persistence. 
Classes: 
Database: obtaining data serialized data stored in .ser files, implements user login functionality, providing User data on login, implements CRUD(create, read update and delete) functionality to allow persistence of any modifications to user data. 

Designs: 
Used interface, UserDataAccessInterface to invert dependency for database access to ensure SOLID design principles. 
Design Patterns: 
•	Unit Tests – 20% only unit tests for reading data, writing data, and login user functionality are tested. Need to implement tests for all functionality in data persistence.
•	Exception Handling – partially complete, need to update taking into consideration use cases. 
•	Java Documentation – incomplete.
•	SOLID Design principles – inverted dependency using dependency inversion principle by creating data access interface.
•	Clean Architecture – implementation in progress

### CODE REVIEW

File: src/main/java/database/Database.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package database;
```
Comment (m-banthwan) [inline src/main/java/database/Database.java:1]: Please add documentation for the methods.

File: src/test/java/test_database/TestDatabaseReadWrite.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import database.Database;
+import entities.*;
+import org.junit.Test;
+
+import java.util.HashMap;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+
+public class TestDatabaseReadWrite {
```
Comment (m-banthwan) [inline src/test/java/test_database/TestDatabaseReadWrite.java:13]: Please add more unit tests for your methods.

File: src/main/java/database/Database.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import java.io.*;
+
+import java.util.HashMap;
+
+public class Database implements UserDataAccessInterface {
+
+    HashMap<String, User> collections;
+    String filePath;
+
+    public Database(String filepath){
```
Comment (happynasit) [inline src/main/java/database/Database.java:15]: Needs improvement in the javadoc

**Review by happynasit — APPROVED** [review]

File: src/test/java/test_database/TestDatabaseReadWrite.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        Timer testTimer = new Timer();
+        User user = new User("test@gmail.com", "test123", "testFirstName", "testLastName", testHabit, testTask, testCat, testTimer);
+        HashMap<String, User> testHashMap = new HashMap();
+        testHashMap.put("test@gmail.com", user);
+        Database.WriteData(testHashMap, "test.ser");
+        Database database = new Database("test.ser");
+        String expected = user.getEmailAddress();
+        String actual = database.LoginUser("test@gmail.com", "test123").getEmailAddress();
+        assertNotEquals(actual, null);
+        assertEquals(actual, expected);
```
Comment (JoAndNight) [inline src/test/java/test_database/TestDatabaseReadWrite.java:30]: Adding an assertion message might be a good idea.

File: src/main/java/user_use_cases/UserDataAccessInterface.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package user_use_cases;
+
+import entities.User;
+
+public interface UserDataAccessInterface {
```
Comment (JoAndNight) [inline src/main/java/user_use_cases/UserDataAccessInterface.java:5]: The use of the Facade design pattern is good. This ensures adhesion to SOLID as well.

**Review by JoAndNight — APPROVED** [review]
Improvements in documentation and tests are needed. The rest is good and can be merged to the main.

**Review by m-banthwan — DISMISSED** [review]

**Review by happynasit — DISMISSED** [review]

**Review by happynasit — APPROVED** [review]

**Pull request merged**